consensus/istanbul/backend/val_enodes_share.go

```diff
@@ -0,0 +1,210 @@
+	return fmt.Sprintf("{Address: %s, EnodeURL: %v, View: %v}", sve.Address.Hex(), sve.EnodeURL, sve.View)
+}
+
+func (sd *valEnodesShareData) String() string {
+	outputStr := "{ValEnodes:"
+	for _, valEnode := range sd.ValEnodes {
+		outputStr = fmt.Sprintf("%s %s", outputStr, valEnode.String())
+	}
+	return fmt.Sprintf("%s}", outputStr)
+}
+
+// ==============================================
+//
+// define the functions that needs to be provided for rlp Encoder/Decoder.
+
+// EncodeRLP serializes sd into the Ethereum RLP format.
+func (sd *valEnodesShareData) EncodeRLP(w io.Writer) error {
+	return rlp.Encode(w, []interface{}{sd.ValEnodes})
+}
+
+// DecodeRLP implements rlp.Decoder, and load the sd fields from a RLP stream.
+func (sd *valEnodesShareData) DecodeRLP(s *rlp.Stream) error {
+	var msg struct {
+		ValEnodes []sharedValidatorEnode
+	}
+
+	if err := s.Decode(&msg); err != nil {
+		return err
+	}
+	sd.ValEnodes = msg.ValEnodes
+	return nil
+}
+
+// This function is meant to be run as a goroutine.  It will periodically send validator enode share messages
+// to this node's proxies so that proxies know the enodes of validators
+func (sb *Backend) sendValEnodesShareMsgs() {
+	sb.valEnodesShareWg.Add(1)
+	defer sb.valEnodesShareWg.Done()
+
+	ticker := time.NewTicker(time.Minute)
+
+	for {
+		select {
+		case <-ticker.C:
+			// output the valEnodeTable for debugging purposes
+			log.Trace("ValidatorEnodeTable dump", "ValidatorEnodeTable", sb.valEnodeTable.String())
+			go sb.sendValEnodesShareMsg()
+
+		case <-sb.valEnodesShareQuit:
+			ticker.Stop()
+			return
+		}
+	}
+}
+
+func (sb *Backend) generateValEnodesShareMsg() (*istanbul.Message, error) {
+	sb.valEnodeTable.valEnodeTableMu.RLock()
+	sharedValidatorEnodes := make([]sharedValidatorEnode, len(sb.valEnodeTable.valEnodeTable))
+	i := 0
+	for address, validatorEnode := range sb.valEnodeTable.valEnodeTable {
+		sharedValidatorEnodes[i] = sharedValidatorEnode{
+			Address:  address,
+			EnodeURL: validatorEnode.node.String(),
+			View:     validatorEnode.view,
+		}
+		i++
+	}
+	sb.valEnodeTable.valEnodeTableMu.RUnlock()
+
+	valEnodesShareData := &valEnodesShareData{
+		ValEnodes: sharedValidatorEnodes,
+	}
+
+	valEnodesShareBytes, err := rlp.EncodeToBytes(valEnodesShareData)
+	if err != nil {
+		sb.logger.Error("Error encoding Istanbul Validator Enodes Share message content", "ValEnodesShareData", valEnodesShareData.String(), "err", err)
+		return nil, err
+	}
+
+	msg := &istanbul.Message{
+		Code:      istanbulValEnodesShareMsg,
+		Msg:       valEnodesShareBytes,
+		Address:   sb.Address(),
+		Signature: []byte{},
+	}
+
+	sb.logger.Trace("Generated a Istanbul Validator Enodes Share message", "IstanbulMsg", msg.String(), "ValEnodesShareData", valEnodesShareData.String())
+
+	return msg, nil
+}
+
+func (sb *Backend) sendValEnodesShareMsg() error {
+	msg, err := sb.generateValEnodesShareMsg()
+	if err != nil {
+		return err
+	}
+
+	if msg == nil {
```

Comment:
When would `msg` ever be nil, but `err` also?

Reply:
good catch